File: pkg/dart2cpp/sample/dart/hello_test_enhanced_simple.dart
/// Dart 简化增强语法测试示例
///
/// 包含适配转换器的更多Dart语法特性
///

void main() {
  print('🔥 Dart 简化增强语法测试开始');

  // 1. 基本数据类型
  testBasicTypes();

  // 2. 控制流
  testControlFlow();

  // 3. 函数（简化版）
  testSimpleFunctions();

  // 4. 集合类型（简化版）
  testSimpleCollections();

  // 5. 类和对象（简化版）
  testSimpleClasses();

  // 6. 泛型（简化版）
  testSimpleGenerics();

  // 7. 异常处理（简化版）
  testSimpleExceptions();

  print('✅ Dart 简化增强语法测试完成');
}

/// 测试基本数据类型
void testBasicTypes() {
  print('\n--- 基本数据类型 ---');

  // 数值类型
  int age = 25;
  double price = 99.99;
  int quantity = 10;

  // 字符串类型
  String name = '张三';
  String message = '你好，世界！';

  // 布尔类型
  bool isStudent = true;
  bool isEmployed = false;

  print('基本类型测试完成');
}

/// 测试控制流
void testControlFlow() {
  print('\n--- 控制流 ---');

  // if-else 语句
  int score = 85;
  if (score >= 90) {
    print('优秀');
  } else if (score >= 80) {
    print('良好');
  } else if (score >= 60) {
    print('及格');
  } else {
    print('不及格');
  }

  // switch 语句
  String grade = 'B';
  switch (grade) {
    case 'A':
      print('优秀');
      break;
    case 'B':
      print('良好');
      break;
    case 'C':
      print('及格');
      break;
    default:
      print('未知等级');
  }

  // for 循环
  print('For 循环:');
  for (int i = 0; i < 5; i++) {
    print('数字: $i');
  }

  // for-in 循环
  List<String> fruits = ['苹果', '香蕉', '橙子'];
  print('For-in 循环:');
  for (String fruit in fruits) {
    print('水果: $fruit');
  }

  // while 循环
  print('While 循环:');
  int count = 3;
  while (count > 0) {
    print('倒计时: $count');
    count--;
  }

  // do-while 循环
  print('Do-while 循环:');
  int number = 1;
  do {
    print('数字: $number');
    number *= 2;
  } while (number <= 10);

  // 条件表达式
  int testAge = 25;
  bool isAdult = testAge >= 18 ? true : false;
  print('成年人: $isAdult');
}

/// 测试简化函数
void testSimpleFunctions() {
  print('\n--- 简化函数 ---');

  // 简单函数
  String greet(String name) {
    return '你好, $name!';
  }

  // 带参数的函数
  int add(int a, int b) {
    return a + b;
  }

  // 调用函数
  print(greet('李四'));
  print('加法结果: ${add(5, 3)}');
}

/// 测试简化集合
void testSimpleCollections() {
  print('\n--- 简化集合 ---');

  // List 操作
  List<int> numbers = [1, 2, 3, 4, 5];
  List<String> names = ['张三', '李四', '王五'];

  // Map 操作
  Map<String, int> scores = {'张三': 95, '李四': 87, '王五': 92};

  // 集合方法
  print('数字列表长度: ${numbers.length}');
  print('第一个数字: ${numbers.first}');
  print('最后一个数字: ${numbers.last}');

  // 列表操作
  numbers.add(6);
  print('添加元素后列表长度: ${numbers.length}');

  // Map 操作
  print('张三的分数: ${scores['张三']}');
  scores['赵六'] = 88;
  print('更新后的分数数量: ${scores.length}');
}

/// 测试简化类
void testSimpleClasses() {
  print('\n--- 简化类 ---');

  // 创建对象
  SimplePerson person = SimplePerson('张三', 25);
  print('创建Person对象');

  // 调用方法
  person.introduce();

  // 访问属性
  print('姓名: ${person.getName()}');
  person.setName('李四');
  print('更新后姓名: ${person.getName()}');
}

/// 测试简化泛型
void testSimpleGenerics() {
  print('\n--- 简化泛型 ---');

  // 泛型类
  SimpleBox<String> stringBox = SimpleBox<String>('Hello');
  SimpleBox<int> intBox = SimpleBox<int>(42);

  print('字符串盒子内容: ${stringBox.getValue()}');
  print('整数盒子内容: ${intBox.getValue()}');
}

/// 测试简化异常处理
void testSimpleExceptions() {
  print('\n--- 简化异常处理 ---');

  try {
    int result = 10 ~/ 2; // 整数除法
    print('结果: $result');
  } catch (e) {
    print('捕获到异常');
  }

  print('异常处理测试完成');
}

// ========================
// 简化类定义
// ========================

/// 简化人员类
class SimplePerson {
  String _name;
  int _age;

  SimplePerson(this._name, this._age);

  // Getter 和 Setter
  String getName() => _name;
  void setName(String value) => _name = value;

  int getAge() => _age;
  void setAge(int value) => _age = value;

  /// 自我介绍
  void introduce() {
    print('大家好，我是 $_name，今年 $_age 岁。');
  }

  @override
  String toString() {
    return 'SimplePerson{name: $_name, age: $_age}';
  }
}

/// 简化泛型盒子类
class SimpleBox<T> {
  T _value;

  SimpleBox(this._value);

  T getValue() => _value;
  void setValue(T value) => _value = value;
}

File: pkg/dart2cpp/sample/cpp_generated/hello_test_enhanced_simple.cpp
#include "dart2cpp.h"

// 工具宏定义

// ============================================================================
// 类: SimplePerson
// ============================================================================

class SimplePerson {
private:
  String _name;
  Int _age;
public:
  SimplePerson(String _name, Int _age) : _name(_name), _age(_age) {
  }
  
  String getName() {
    return this->_name;
  }
  
  String setName(String value) {
    return this->_name = value;
  }
  
  Int getAge() {
    return this->_age;
  }
  
  Int setAge(Int value) {
    return this->_age = value;
  }
  
  Nullable introduce() {
    dart_print(dart_concat(dart_string("大家好，我是 "), (this->_name).toString(), dart_string("，今年 "), (this->_age).toString(), dart_string(" 岁。")));
return Void;
  }
  
  String toString() {
    return dart_concat(dart_string("SimplePerson{name: "), (this->_name).toString(), dart_string(", age: "), (this->_age).toString(), dart_string("}"));
  }
  
};

// ============================================================================
// 类: SimpleBox
// ============================================================================

template<typename T>
class SimpleBox {
private:
  T _value;
public:
  SimpleBox(T _value) : _value(_value) {
  }
  
  T getValue() {
    return this->_value;
  }
  
  T setValue(T value) {
    return this->_value = value;
  }
  
};

Nullable testBasicTypes();
Nullable testControlFlow();
Nullable testSimpleFunctions();
Nullable testSimpleCollections();
Nullable testSimpleClasses();
Nullable testSimpleGenerics();
Nullable testSimpleExceptions();
Nullable testBasicTypes() {
  dart_print(dart_string("\n--- 基本数据类型 ---"));
auto age = dart_int(25);
auto price = dart_double(99.99);
auto quantity = dart_int(10);
auto name = dart_string("张三");
auto message = dart_string("你好，世界！");
auto isStudent = dart_bool(true);
auto isEmployed = dart_bool(false);
dart_print(dart_string("基本类型测试完成"));
return Void;
}

Nullable testControlFlow() {
  dart_print(dart_string("\n--- 控制流 ---"));
auto score = dart_int(85);
if (score->operator_greater_equals(dart_int(90))) {
dart_print(dart_string("优秀"));
} else {
if (score->operator_greater_equals(dart_int(80))) {
dart_print(dart_string("良好"));
} else {
if (score->operator_greater_equals(dart_int(60))) {
dart_print(dart_string("及格"));
} else {
dart_print(dart_string("不及格"));
}
}
}
auto grade = dart_string("B");
if (grade == dart_string("A")) {
  dart_print(dart_string("优秀"));
} else if (grade == dart_string("B")) {
  dart_print(dart_string("良好"));
} else if (grade == dart_string("C")) {
  dart_print(dart_string("及格"));
}else {
  dart_print(dart_string("未知等级"));
}
dart_print(dart_string("For 循环:"));
for (auto i = dart_int(0); i->operator_less(dart_int(5)); i = i->operator_add(dart_int(1))) {
dart_print(dart_string("数字: ") + (i).toString());
}
auto fruits = dart_literal<String>(dart_string("苹果"), dart_string("香蕉"), dart_string("橙子"));
dart_print(dart_string("For-in 循环:"));
auto sync_for_iterator = fruits->iterator();
for (; sync_for_iterator->hasNext(); ) {
auto fruit = sync_for_iterator->next();
dart_print(dart_string("水果: ") + (fruit).toString());
}
dart_print(dart_string("While 循环:"));
auto count = dart_int(3);
while (count->operator_greater(dart_int(0))) {
dart_print(dart_string("倒计时: ") + (count).toString());
count = count->operator_sub(dart_int(1));
}
dart_print(dart_string("Do-while 循环:"));
auto number = dart_int(1);
do {
dart_print(dart_string("数字: ") + (number).toString());
number = number->operator_mul(dart_int(2));
} while ((number->operator_less_equals(dart_int(10))).toBool());
auto testAge = dart_int(25);
auto isAdult = testAge->operator_greater_equals(dart_int(18)) ? dart_bool(true) : dart_bool(false);
dart_print(dart_string("成年人: ") + (isAdult).toString());
return Void;
}

Nullable testSimpleFunctions() {
  dart_print(dart_string("\n--- 简化函数 ---"));
auto greet = [&](String name) { return dart_concat(dart_string("你好, "), (name).toString(), dart_string("!")); };
auto add = [&](Int a, Int b) { return a->operator_add(b); };
dart_print(greet(dart_string("李四")));
dart_print(dart_string("加法结果: ") + (add(dart_int(5), dart_int(3))).toString());
return Void;
}

Nullable testSimpleCollections() {
  dart_print(dart_string("\n--- 简化集合 ---"));
auto numbers = dart_literal<Int>(dart_int(1), dart_int(2), dart_int(3), dart_int(4), dart_int(5));
auto names = dart_literal<String>(dart_string("张三"), dart_string("李四"), dart_string("王五"));
auto scores = Map<String, Int>::createFromEntries({{dart_string("张三"), dart_int(95)}, {dart_string("李四"), dart_int(87)}, {dart_string("王五"), dart_int(92)}});
dart_print(dart_string("数字列表长度: ") + (numbers->size()).toString());
dart_print(dart_string("第一个数字: ") + (numbers->first()).toString());
dart_print(dart_string("最后一个数字: ") + (numbers->last()).toString());
numbers->add(dart_int(6));
dart_print(dart_string("添加元素后列表长度: ") + (numbers->size()).toString());
dart_print(dart_string("张三的分数: ") + (scores->operator_index(dart_string("张三"))).toString());
scores->operator_index_set(dart_string("赵六"), dart_int(88));
dart_print(dart_string("更新后的分数数量: ") + (scores->size()).toString());
return Void;
}

Nullable testSimpleClasses() {
  dart_print(dart_string("\n--- 简化类 ---"));
auto person = ObjectPtr<SimplePerson>(new SimplePerson(dart_string("张三"), dart_int(25)));
dart_print(dart_string("创建Person对象"));
person->introduce();
dart_print(dart_string("姓名: ") + (person->getName()).toString());
person->setName(dart_string("李四"));
dart_print(dart_string("更新后姓名: ") + (person->getName()).toString());
return Void;
}

Nullable testSimpleGenerics() {
  dart_print(dart_string("\n--- 简化泛型 ---"));
auto stringBox = ObjectPtr<SimpleBox<String>>(new SimpleBox<String>(dart_string("Hello")));
auto intBox = ObjectPtr<SimpleBox<Int>>(new SimpleBox<Int>(dart_int(42)));
dart_print(dart_string("字符串盒子内容: ") + (stringBox->getValue()).toString());
dart_print(dart_string("整数盒子内容: ") + (intBox->getValue()).toString());
return Void;
}

Nullable testSimpleExceptions() {
  dart_print(dart_string("\n--- 简化异常处理 ---"));
try {
auto result = dart_int(10)->truncatingDivision(dart_int(2));
dart_print(dart_string("结果: ") + (result).toString());
} catch (const std::exception& e) { /* catch block */ }
// Finally block should be implemented using RAII pattern
dart_print(dart_string("异常处理测试完成"));
return Void;
}

// ============================================================================
// 主函数
// ============================================================================

int main() {
  try {
    dart_print(dart_string("🔥 Dart 简化增强语法测试开始"));
testBasicTypes();
testControlFlow();
testSimpleFunctions();
testSimpleCollections();
testSimpleClasses();
testSimpleGenerics();
testSimpleExceptions();
dart_print(dart_string("✅ Dart 简化增强语法测试完成"));
    return 0;
  } catch (const std::exception& e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}
